datafusion/core/src/execution/session_state.rs

```diff
@@ -976,6 +872,713 @@ impl SessionState {
+    optimizer_rules: Option<Vec<Arc<dyn OptimizerRule + Send + Sync>>>,
+    physical_optimizer_rules: Option<Vec<Arc<dyn PhysicalOptimizerRule + Send + Sync>>>,
+}
+
+impl SessionStateBuilder {
+    /// Returns a new [`SessionStateBuilder`] with no options set.
+    pub fn new() -> Self {
+        Self {
+            session_id: None,
+            analyzer: None,
+            expr_planners: None,
+            optimizer: None,
+            physical_optimizers: None,
+            query_planner: None,
+            catalog_list: None,
+            table_functions: None,
+            scalar_functions: None,
+            aggregate_functions: None,
+            window_functions: None,
+            serializer_registry: None,
+            file_formats: None,
+            table_options: None,
+            config: None,
+            execution_props: None,
+            table_factories: None,
+            runtime_env: None,
+            function_factory: None,
+            // fields to support convenience functions
+            analyzer_rules: None,
+            optimizer_rules: None,
+            physical_optimizer_rules: None,
+        }
+    }
+
+    /// Returns a new [SessionStateBuilder] based on an existing [SessionState]
+    /// The session id for the new builder will be unset; all other fields will
+    /// be cloned from what is set in the provided session state
+    pub fn new_from_existing(existing: SessionState) -> Self {
+        let cloned = existing.clone();
+
+        Self {
+            session_id: None,
+            analyzer: Some(cloned.analyzer),
+            expr_planners: Some(cloned.expr_planners),
+            optimizer: Some(cloned.optimizer),
+            physical_optimizers: Some(cloned.physical_optimizers),
+            query_planner: Some(cloned.query_planner),
+            catalog_list: Some(cloned.catalog_list),
+            table_functions: Some(cloned.table_functions),
+            scalar_functions: Some(cloned.scalar_functions.into_values().collect_vec()),
+            aggregate_functions: Some(
+                cloned.aggregate_functions.into_values().collect_vec(),
+            ),
+            window_functions: Some(cloned.window_functions.into_values().collect_vec()),
+            serializer_registry: Some(cloned.serializer_registry),
+            file_formats: Some(cloned.file_formats.into_values().collect_vec()),
+            config: Some(cloned.config),
+            table_options: Some(cloned.table_options),
+            execution_props: Some(cloned.execution_props),
+            table_factories: Some(cloned.table_factories),
+            runtime_env: Some(cloned.runtime_env),
+            function_factory: cloned.function_factory,
+
+            // fields to support convenience functions
+            analyzer_rules: None,
+            optimizer_rules: None,
+            physical_optimizer_rules: None,
+        }
+    }
+
+    /// Set defaults for table_factories, file formats, expr_planners and builtin
```

Comment:
I think this is great for this PR

As part of the larger catalog refactor, I think we'll ned to figure out how to get SessioNStateDefaults` out of this file -- but I have some ideas (specifically we can make the defaults a trait that could be passed to the builder).

Not for this PR